File: clients/kmp/shared/src/commonMain/kotlin/org/wyrdsekai/app/engine/room/RoomMemoryPolicy.kt
package org.wyrdsekai.app.engine.room

import org.wyrdsekai.app.engine.event.WorldEvent

/**
 * Three-tier conversation buffer for room memory management.
 * Port of core/room/RoomMemoryPolicy.java.
 *
 * Hot: last N messages (full detail, high priority for prompts)
 * Warm: next M messages (available but lower priority)
 * Compacted: oldest messages, summarized or dropped
 */
class RoomMemoryPolicy(
    private val hotSize: Int = 10,
    private val warmSize: Int = 20,
) {
    private val _hotEvents = mutableListOf<WorldEvent.Said>()
    private val _warmEvents = mutableListOf<WorldEvent.Said>()

    fun add(event: WorldEvent.Said) {
        _hotEvents.add(event)
        // Cascade: hot overflows to warm, warm overflows to compacted (dropped)
        while (_hotEvents.size > hotSize) {
            val overflow = _hotEvents.removeFirst()
            _warmEvents.add(overflow)
        }
        while (_warmEvents.size > warmSize) {
            _warmEvents.removeFirst() // compacted = dropped for phone node
        }
    }

    /** Recent messages for prompt assembly. */
    fun hotEvents(): List<WorldEvent.Said> = _hotEvents.toList()

    /** Older messages available for expanded context. */
    fun warmEvents(): List<WorldEvent.Said> = _warmEvents.toList()

    /** All retained messages (hot + warm). */
    fun allEvents(): List<WorldEvent.Said> = _warmEvents + _hotEvents

    fun clear() {
        _hotEvents.clear()
        _warmEvents.clear()
    }

    companion object {
        fun default() = RoomMemoryPolicy()
    }
}

File: clients/rn/src/engine/room/RoomMemoryPolicy.ts
/**
 * Three-tier conversation buffer for room memory management.
 * TypeScript port of KMP's RoomMemoryPolicy.kt.
 *
 * Hot: last N messages (full detail, high priority for prompts)
 * Warm: next M messages (available but lower priority)
 * Compacted: oldest messages, summarized or dropped
 */

import type { Said } from '../events/WorldEvent';

export class RoomMemoryPolicy {
  private hotEvents: Said[] = [];
  private warmEvents: Said[] = [];

  constructor(
    private readonly hotSize: number = 10,
    private readonly warmSize: number = 20,
  ) {}

  add(event: Said): void {
    this.hotEvents.push(event);
    // Cascade: hot overflows to warm, warm overflows to compacted (dropped)
    while (this.hotEvents.length > this.hotSize) {
      const overflow = this.hotEvents.shift()!;
      this.warmEvents.push(overflow);
    }
    while (this.warmEvents.length > this.warmSize) {
      this.warmEvents.shift(); // compacted = dropped for phone node
    }
  }

  /** Recent messages for prompt assembly. */
  getHotEvents(): Said[] {
    return [...this.hotEvents];
  }

  /** Older messages available for expanded context. */
  getWarmEvents(): Said[] {
    return [...this.warmEvents];
  }

  /** All retained messages (hot + warm). */
  getAllEvents(): Said[] {
    return [...this.warmEvents, ...this.hotEvents];
  }

  clear(): void {
    this.hotEvents = [];
    this.warmEvents = [];
  }

  static default(): RoomMemoryPolicy {
    return new RoomMemoryPolicy();
  }
}
